# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [17]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    def process_sequence(txt, dct, is_t = False):
        r = []
        for l in txt.lower().split('\n'):
            c = []
            for w in l.split():
                c.append(dct[w])
            if is_t:
                c.append(dct['<EOS>'])
            r.append(c)
        return r
    r_1 = process_sequence(source_text, source_vocab_to_int)
    r_2 = process_sequence(target_text, target_vocab_to_int, True)
   
    return r_1, r_2

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [3]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    input_data = tf.placeholder(tf.int32, [None, None], name='input_data')
    targets = tf.placeholder(tf.int32, [None, None], name='targets')
    learning_rate = tf.placeholder(tf.float32, name='learning_rate')
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None, ], name='target_sequence_length')
    max_target_sequence_length = tf.reduce_max(target_sequence_length, name='max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, [None, ], name='source_sequence_length')
    return input_data, targets, learning_rate, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "c:\\anaconda3\\envs\\udacity_learning_2_35\\lib\\runpy.py", line 193, in _run_module_as_main\n    "__main__", mod_spec)', 'File "c:\\anaconda3\\envs\\udacity_learning_2_35\\lib\\runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "c:\\anaconda3\\envs\\udacity_learning_2_35\\lib\\site-packages\\ipykernel_launcher.py", line 16, in <module>\n    app.launch_new_instance()', 'File "c:\\anaconda3\\envs\\udacity_learning_2_35\\lib\\site-packages\\traitlets\\config\\application.py", line 658, in launch_instance\n    app.start()', 'File "c:\\anaconda3\\envs\\udacity_learning_2_35\\lib\\site-packages\\ipykernel\\kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [4]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [5]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size)
        drop = tf.contrib.rnn.DropoutWrapper(enc_cell, output_keep_prob=keep_prob)
        return drop
    # Encoder embedding
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    
    # Stack up multiple LSTM layers, for deep learning
    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    enc_output, enc_state = tf.nn.dynamic_rnn(
        enc_cell, 
        enc_embed_input, 
        sequence_length=source_sequence_length, 
        dtype=tf.float32
    )
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [6]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    drop =  tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input, sequence_length=target_sequence_length, time_major=False)
    training_decoder = tf.contrib.seq2seq.BasicDecoder(drop, training_helper, encoder_state, output_layer)
    training_decoder_output, _, __ = tf.contrib.seq2seq.dynamic_decode(training_decoder, impute_finished=True, maximum_iterations=max_summary_length)
    
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [7]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state !
    :param dec_cell: Decoder RNN Cell !
    :param dec_embeddings: Decoder embeddings !
    :param start_of_sequence_id: GO ID !
    :param end_of_sequence_id: EOS Id !
    :param max_target_sequence_length: Maximum length of target sequences !
    :param vocab_size: Size of decoder/target vocabulary
    :param output_layer: Function to apply the output layer !
    :param batch_size: Batch size !
    :param keep_prob: Dropout keep probability !
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    drop =  tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)

    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(drop, inference_helper, encoder_state, output_layer)
    inference_decoder_output, _, __ = tf.contrib.seq2seq.dynamic_decode(
        inference_decoder,
        impute_finished=True,
        maximum_iterations=max_target_sequence_length
    )
    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [8]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size !
    :param num_layers: Number of layers !
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size)
        return dec_cell
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    output_layer = Dense(target_vocab_size, kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope("decode"):
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)
    with tf.variable_scope("decode", reuse=True):
        inference_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], max_target_sequence_length, target_vocab_size, output_layer, batch_size, keep_prob) 
    
    return training_decoder_output, inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [9]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    _, encoder_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   enc_embedding_size)
    
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sentence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, dec_embedding_size)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [10]:
# Number of Epochs
epochs = 60
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 50
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 15
decoding_embedding_size = 15
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.5
display_step = 50

### Build the Graph
Build the graph using the neural network you implemented.

In [11]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [12]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [13]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   50/1077 - Train Accuracy: 0.2602, Validation Accuracy: 0.3363, Loss: 3.7936
Epoch   0 Batch  100/1077 - Train Accuracy: 0.2645, Validation Accuracy: 0.3377, Loss: 3.3932
Epoch   0 Batch  150/1077 - Train Accuracy: 0.3616, Validation Accuracy: 0.3768, Loss: 2.9770
Epoch   0 Batch  200/1077 - Train Accuracy: 0.3523, Validation Accuracy: 0.4031, Loss: 2.9329
Epoch   0 Batch  250/1077 - Train Accuracy: 0.4087, Validation Accuracy: 0.4229, Loss: 2.6439
Epoch   0 Batch  300/1077 - Train Accuracy: 0.3376, Validation Accuracy: 0.4318, Loss: 2.8984
Epoch   0 Batch  350/1077 - Train Accuracy: 0.3680, Validation Accuracy: 0.4418, Loss: 2.7400
Epoch   0 Batch  400/1077 - Train Accuracy: 0.3973, Validation Accuracy: 0.4627, Loss: 2.5921
Epoch   0 Batch  450/1077 - Train Accuracy: 0.4258, Validation Accuracy: 0.4759, Loss: 2.4980
Epoch   0 Batch  500/1077 - Train Accuracy: 0.4316, Validation Accuracy: 0.4780, Loss: 2.3830
Epoch   0 Batch  550/1077 - Train Accuracy: 0.4191, Validati

Epoch   4 Batch  250/1077 - Train Accuracy: 0.5781, Validation Accuracy: 0.6193, Loss: 0.6943
Epoch   4 Batch  300/1077 - Train Accuracy: 0.6106, Validation Accuracy: 0.6218, Loss: 0.7633
Epoch   4 Batch  350/1077 - Train Accuracy: 0.5832, Validation Accuracy: 0.6261, Loss: 0.7635
Epoch   4 Batch  400/1077 - Train Accuracy: 0.6117, Validation Accuracy: 0.6360, Loss: 0.7390
Epoch   4 Batch  450/1077 - Train Accuracy: 0.6031, Validation Accuracy: 0.6325, Loss: 0.7070
Epoch   4 Batch  500/1077 - Train Accuracy: 0.6238, Validation Accuracy: 0.6264, Loss: 0.7231
Epoch   4 Batch  550/1077 - Train Accuracy: 0.5852, Validation Accuracy: 0.6307, Loss: 0.7569
Epoch   4 Batch  600/1077 - Train Accuracy: 0.6254, Validation Accuracy: 0.6257, Loss: 0.6758
Epoch   4 Batch  650/1077 - Train Accuracy: 0.5992, Validation Accuracy: 0.6246, Loss: 0.7173
Epoch   4 Batch  700/1077 - Train Accuracy: 0.6023, Validation Accuracy: 0.6286, Loss: 0.6858
Epoch   4 Batch  750/1077 - Train Accuracy: 0.6160, Validati

Epoch   8 Batch  450/1077 - Train Accuracy: 0.6766, Validation Accuracy: 0.7035, Loss: 0.4753
Epoch   8 Batch  500/1077 - Train Accuracy: 0.7078, Validation Accuracy: 0.7184, Loss: 0.4897
Epoch   8 Batch  550/1077 - Train Accuracy: 0.6730, Validation Accuracy: 0.6953, Loss: 0.5064
Epoch   8 Batch  600/1077 - Train Accuracy: 0.7228, Validation Accuracy: 0.7038, Loss: 0.4568
Epoch   8 Batch  650/1077 - Train Accuracy: 0.6926, Validation Accuracy: 0.7148, Loss: 0.5067
Epoch   8 Batch  700/1077 - Train Accuracy: 0.6687, Validation Accuracy: 0.7085, Loss: 0.4744
Epoch   8 Batch  750/1077 - Train Accuracy: 0.7156, Validation Accuracy: 0.7099, Loss: 0.4805
Epoch   8 Batch  800/1077 - Train Accuracy: 0.6758, Validation Accuracy: 0.7131, Loss: 0.4872
Epoch   8 Batch  850/1077 - Train Accuracy: 0.6629, Validation Accuracy: 0.7202, Loss: 0.5067
Epoch   8 Batch  900/1077 - Train Accuracy: 0.6973, Validation Accuracy: 0.7035, Loss: 0.4769
Epoch   8 Batch  950/1077 - Train Accuracy: 0.6570, Validati

Epoch  12 Batch  650/1077 - Train Accuracy: 0.7586, Validation Accuracy: 0.7724, Loss: 0.3691
Epoch  12 Batch  700/1077 - Train Accuracy: 0.7695, Validation Accuracy: 0.7571, Loss: 0.3545
Epoch  12 Batch  750/1077 - Train Accuracy: 0.7637, Validation Accuracy: 0.7521, Loss: 0.3859
Epoch  12 Batch  800/1077 - Train Accuracy: 0.7750, Validation Accuracy: 0.7834, Loss: 0.3704
Epoch  12 Batch  850/1077 - Train Accuracy: 0.7251, Validation Accuracy: 0.7805, Loss: 0.4111
Epoch  12 Batch  900/1077 - Train Accuracy: 0.7609, Validation Accuracy: 0.7678, Loss: 0.3745
Epoch  12 Batch  950/1077 - Train Accuracy: 0.7288, Validation Accuracy: 0.7749, Loss: 0.3614
Epoch  12 Batch 1000/1077 - Train Accuracy: 0.7887, Validation Accuracy: 0.7827, Loss: 0.3408
Epoch  12 Batch 1050/1077 - Train Accuracy: 0.7305, Validation Accuracy: 0.7734, Loss: 0.3872
Epoch  13 Batch   50/1077 - Train Accuracy: 0.7727, Validation Accuracy: 0.7734, Loss: 0.3755
Epoch  13 Batch  100/1077 - Train Accuracy: 0.7594, Validati

Epoch  16 Batch  850/1077 - Train Accuracy: 0.7515, Validation Accuracy: 0.7844, Loss: 0.3470
Epoch  16 Batch  900/1077 - Train Accuracy: 0.8168, Validation Accuracy: 0.7891, Loss: 0.3196
Epoch  16 Batch  950/1077 - Train Accuracy: 0.7333, Validation Accuracy: 0.7763, Loss: 0.2845
Epoch  16 Batch 1000/1077 - Train Accuracy: 0.8069, Validation Accuracy: 0.7827, Loss: 0.2638
Epoch  16 Batch 1050/1077 - Train Accuracy: 0.7414, Validation Accuracy: 0.7933, Loss: 0.3161
Epoch  17 Batch   50/1077 - Train Accuracy: 0.7977, Validation Accuracy: 0.7979, Loss: 0.3051
Epoch  17 Batch  100/1077 - Train Accuracy: 0.7793, Validation Accuracy: 0.7947, Loss: 0.2955
Epoch  17 Batch  150/1077 - Train Accuracy: 0.8103, Validation Accuracy: 0.7951, Loss: 0.3000
Epoch  17 Batch  200/1077 - Train Accuracy: 0.7766, Validation Accuracy: 0.8068, Loss: 0.3044
Epoch  17 Batch  250/1077 - Train Accuracy: 0.8026, Validation Accuracy: 0.7951, Loss: 0.2861
Epoch  17 Batch  300/1077 - Train Accuracy: 0.8261, Validati

Epoch  20 Batch 1050/1077 - Train Accuracy: 0.8109, Validation Accuracy: 0.8267, Loss: 0.2541
Epoch  21 Batch   50/1077 - Train Accuracy: 0.8746, Validation Accuracy: 0.8278, Loss: 0.2599
Epoch  21 Batch  100/1077 - Train Accuracy: 0.8207, Validation Accuracy: 0.8523, Loss: 0.2419
Epoch  21 Batch  150/1077 - Train Accuracy: 0.8672, Validation Accuracy: 0.8200, Loss: 0.2428
Epoch  21 Batch  200/1077 - Train Accuracy: 0.8219, Validation Accuracy: 0.8469, Loss: 0.2618
Epoch  21 Batch  250/1077 - Train Accuracy: 0.8391, Validation Accuracy: 0.8299, Loss: 0.2390
Epoch  21 Batch  300/1077 - Train Accuracy: 0.8914, Validation Accuracy: 0.8512, Loss: 0.2485
Epoch  21 Batch  350/1077 - Train Accuracy: 0.8883, Validation Accuracy: 0.8359, Loss: 0.2444
Epoch  21 Batch  400/1077 - Train Accuracy: 0.8543, Validation Accuracy: 0.8303, Loss: 0.2675
Epoch  21 Batch  450/1077 - Train Accuracy: 0.8508, Validation Accuracy: 0.8068, Loss: 0.2383
Epoch  21 Batch  500/1077 - Train Accuracy: 0.8633, Validati

Epoch  25 Batch  200/1077 - Train Accuracy: 0.8449, Validation Accuracy: 0.8594, Loss: 0.2249
Epoch  25 Batch  250/1077 - Train Accuracy: 0.8771, Validation Accuracy: 0.8409, Loss: 0.2004
Epoch  25 Batch  300/1077 - Train Accuracy: 0.9248, Validation Accuracy: 0.8732, Loss: 0.2246
Epoch  25 Batch  350/1077 - Train Accuracy: 0.8844, Validation Accuracy: 0.8501, Loss: 0.2134
Epoch  25 Batch  400/1077 - Train Accuracy: 0.8840, Validation Accuracy: 0.8711, Loss: 0.2329
Epoch  25 Batch  450/1077 - Train Accuracy: 0.8883, Validation Accuracy: 0.8544, Loss: 0.2304
Epoch  25 Batch  500/1077 - Train Accuracy: 0.8895, Validation Accuracy: 0.8690, Loss: 0.1976
Epoch  25 Batch  550/1077 - Train Accuracy: 0.8313, Validation Accuracy: 0.8647, Loss: 0.2313
Epoch  25 Batch  600/1077 - Train Accuracy: 0.8791, Validation Accuracy: 0.8473, Loss: 0.2025
Epoch  25 Batch  650/1077 - Train Accuracy: 0.8734, Validation Accuracy: 0.8587, Loss: 0.2410
Epoch  25 Batch  700/1077 - Train Accuracy: 0.8949, Validati

Epoch  29 Batch  400/1077 - Train Accuracy: 0.8938, Validation Accuracy: 0.8881, Loss: 0.2297
Epoch  29 Batch  450/1077 - Train Accuracy: 0.9027, Validation Accuracy: 0.8697, Loss: 0.1900
Epoch  29 Batch  500/1077 - Train Accuracy: 0.9078, Validation Accuracy: 0.8899, Loss: 0.1732
Epoch  29 Batch  550/1077 - Train Accuracy: 0.8527, Validation Accuracy: 0.8871, Loss: 0.1982
Epoch  29 Batch  600/1077 - Train Accuracy: 0.8969, Validation Accuracy: 0.8725, Loss: 0.2011
Epoch  29 Batch  650/1077 - Train Accuracy: 0.9051, Validation Accuracy: 0.8764, Loss: 0.2005
Epoch  29 Batch  700/1077 - Train Accuracy: 0.8965, Validation Accuracy: 0.8690, Loss: 0.1729
Epoch  29 Batch  750/1077 - Train Accuracy: 0.8953, Validation Accuracy: 0.8832, Loss: 0.2008
Epoch  29 Batch  800/1077 - Train Accuracy: 0.9230, Validation Accuracy: 0.8711, Loss: 0.1904
Epoch  29 Batch  850/1077 - Train Accuracy: 0.8847, Validation Accuracy: 0.8888, Loss: 0.2128
Epoch  29 Batch  900/1077 - Train Accuracy: 0.9000, Validati

Epoch  33 Batch  600/1077 - Train Accuracy: 0.9118, Validation Accuracy: 0.8917, Loss: 0.1701
Epoch  33 Batch  650/1077 - Train Accuracy: 0.9105, Validation Accuracy: 0.8878, Loss: 0.1832
Epoch  33 Batch  700/1077 - Train Accuracy: 0.9301, Validation Accuracy: 0.8938, Loss: 0.1570
Epoch  33 Batch  750/1077 - Train Accuracy: 0.8910, Validation Accuracy: 0.8931, Loss: 0.1922
Epoch  33 Batch  800/1077 - Train Accuracy: 0.9195, Validation Accuracy: 0.9038, Loss: 0.1602
Epoch  33 Batch  850/1077 - Train Accuracy: 0.9014, Validation Accuracy: 0.9020, Loss: 0.1960
Epoch  33 Batch  900/1077 - Train Accuracy: 0.9027, Validation Accuracy: 0.8977, Loss: 0.1873
Epoch  33 Batch  950/1077 - Train Accuracy: 0.9040, Validation Accuracy: 0.8952, Loss: 0.1547
Epoch  33 Batch 1000/1077 - Train Accuracy: 0.9133, Validation Accuracy: 0.9031, Loss: 0.1561
Epoch  33 Batch 1050/1077 - Train Accuracy: 0.8961, Validation Accuracy: 0.8970, Loss: 0.1724
Epoch  34 Batch   50/1077 - Train Accuracy: 0.9113, Validati

Epoch  37 Batch  800/1077 - Train Accuracy: 0.9422, Validation Accuracy: 0.8981, Loss: 0.1617
Epoch  37 Batch  850/1077 - Train Accuracy: 0.9014, Validation Accuracy: 0.9137, Loss: 0.1917
Epoch  37 Batch  900/1077 - Train Accuracy: 0.9164, Validation Accuracy: 0.9062, Loss: 0.1717
Epoch  37 Batch  950/1077 - Train Accuracy: 0.9152, Validation Accuracy: 0.9038, Loss: 0.1491
Epoch  37 Batch 1000/1077 - Train Accuracy: 0.9100, Validation Accuracy: 0.9048, Loss: 0.1417
Epoch  37 Batch 1050/1077 - Train Accuracy: 0.9211, Validation Accuracy: 0.9027, Loss: 0.1602
Epoch  38 Batch   50/1077 - Train Accuracy: 0.9391, Validation Accuracy: 0.8924, Loss: 0.1709
Epoch  38 Batch  100/1077 - Train Accuracy: 0.9156, Validation Accuracy: 0.9073, Loss: 0.1618
Epoch  38 Batch  150/1077 - Train Accuracy: 0.9144, Validation Accuracy: 0.8960, Loss: 0.1574
Epoch  38 Batch  200/1077 - Train Accuracy: 0.9148, Validation Accuracy: 0.8984, Loss: 0.1632
Epoch  38 Batch  250/1077 - Train Accuracy: 0.8963, Validati

Epoch  41 Batch 1000/1077 - Train Accuracy: 0.9036, Validation Accuracy: 0.8952, Loss: 0.1464
Epoch  41 Batch 1050/1077 - Train Accuracy: 0.9219, Validation Accuracy: 0.9073, Loss: 0.1536
Epoch  42 Batch   50/1077 - Train Accuracy: 0.9191, Validation Accuracy: 0.9055, Loss: 0.1536
Epoch  42 Batch  100/1077 - Train Accuracy: 0.9031, Validation Accuracy: 0.8931, Loss: 0.1527
Epoch  42 Batch  150/1077 - Train Accuracy: 0.9219, Validation Accuracy: 0.9126, Loss: 0.1508
Epoch  42 Batch  200/1077 - Train Accuracy: 0.9187, Validation Accuracy: 0.9031, Loss: 0.1568
Epoch  42 Batch  250/1077 - Train Accuracy: 0.9215, Validation Accuracy: 0.9006, Loss: 0.1412
Epoch  42 Batch  300/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9055, Loss: 0.1478
Epoch  42 Batch  350/1077 - Train Accuracy: 0.9211, Validation Accuracy: 0.8867, Loss: 0.1551
Epoch  42 Batch  400/1077 - Train Accuracy: 0.9062, Validation Accuracy: 0.9041, Loss: 0.1767
Epoch  42 Batch  450/1077 - Train Accuracy: 0.9254, Validati

Epoch  46 Batch  150/1077 - Train Accuracy: 0.9018, Validation Accuracy: 0.8981, Loss: 0.1340
Epoch  46 Batch  200/1077 - Train Accuracy: 0.9098, Validation Accuracy: 0.9016, Loss: 0.1414
Epoch  46 Batch  250/1077 - Train Accuracy: 0.9322, Validation Accuracy: 0.9197, Loss: 0.1315
Epoch  46 Batch  300/1077 - Train Accuracy: 0.9704, Validation Accuracy: 0.9226, Loss: 0.1404
Epoch  46 Batch  350/1077 - Train Accuracy: 0.9262, Validation Accuracy: 0.9002, Loss: 0.1311
Epoch  46 Batch  400/1077 - Train Accuracy: 0.9238, Validation Accuracy: 0.9105, Loss: 0.1603
Epoch  46 Batch  450/1077 - Train Accuracy: 0.9480, Validation Accuracy: 0.9070, Loss: 0.1466
Epoch  46 Batch  500/1077 - Train Accuracy: 0.9258, Validation Accuracy: 0.9109, Loss: 0.1220
Epoch  46 Batch  550/1077 - Train Accuracy: 0.9012, Validation Accuracy: 0.9098, Loss: 0.1410
Epoch  46 Batch  600/1077 - Train Accuracy: 0.9249, Validation Accuracy: 0.9173, Loss: 0.1413
Epoch  46 Batch  650/1077 - Train Accuracy: 0.9262, Validati

Epoch  50 Batch  350/1077 - Train Accuracy: 0.9523, Validation Accuracy: 0.9105, Loss: 0.1212
Epoch  50 Batch  400/1077 - Train Accuracy: 0.9203, Validation Accuracy: 0.9130, Loss: 0.1540
Epoch  50 Batch  450/1077 - Train Accuracy: 0.9422, Validation Accuracy: 0.9038, Loss: 0.1381
Epoch  50 Batch  500/1077 - Train Accuracy: 0.9285, Validation Accuracy: 0.9187, Loss: 0.1057
Epoch  50 Batch  550/1077 - Train Accuracy: 0.8945, Validation Accuracy: 0.9077, Loss: 0.1448
Epoch  50 Batch  600/1077 - Train Accuracy: 0.9315, Validation Accuracy: 0.9251, Loss: 0.1282
Epoch  50 Batch  650/1077 - Train Accuracy: 0.9410, Validation Accuracy: 0.9208, Loss: 0.1426
Epoch  50 Batch  700/1077 - Train Accuracy: 0.9520, Validation Accuracy: 0.9066, Loss: 0.1188
Epoch  50 Batch  750/1077 - Train Accuracy: 0.9316, Validation Accuracy: 0.9165, Loss: 0.1428
Epoch  50 Batch  800/1077 - Train Accuracy: 0.9383, Validation Accuracy: 0.9290, Loss: 0.1262
Epoch  50 Batch  850/1077 - Train Accuracy: 0.9036, Validati

Epoch  54 Batch  550/1077 - Train Accuracy: 0.8977, Validation Accuracy: 0.9162, Loss: 0.1237
Epoch  54 Batch  600/1077 - Train Accuracy: 0.9435, Validation Accuracy: 0.9237, Loss: 0.1288
Epoch  54 Batch  650/1077 - Train Accuracy: 0.9434, Validation Accuracy: 0.9087, Loss: 0.1382
Epoch  54 Batch  700/1077 - Train Accuracy: 0.9449, Validation Accuracy: 0.9052, Loss: 0.1155
Epoch  54 Batch  750/1077 - Train Accuracy: 0.9336, Validation Accuracy: 0.9134, Loss: 0.1444
Epoch  54 Batch  800/1077 - Train Accuracy: 0.9410, Validation Accuracy: 0.9173, Loss: 0.1147
Epoch  54 Batch  850/1077 - Train Accuracy: 0.9066, Validation Accuracy: 0.9123, Loss: 0.1504
Epoch  54 Batch  900/1077 - Train Accuracy: 0.9305, Validation Accuracy: 0.9308, Loss: 0.1378
Epoch  54 Batch  950/1077 - Train Accuracy: 0.9289, Validation Accuracy: 0.9183, Loss: 0.1260
Epoch  54 Batch 1000/1077 - Train Accuracy: 0.9170, Validation Accuracy: 0.9059, Loss: 0.1168
Epoch  54 Batch 1050/1077 - Train Accuracy: 0.9426, Validati

Epoch  58 Batch  750/1077 - Train Accuracy: 0.9336, Validation Accuracy: 0.9141, Loss: 0.1239
Epoch  58 Batch  800/1077 - Train Accuracy: 0.9414, Validation Accuracy: 0.9151, Loss: 0.1061
Epoch  58 Batch  850/1077 - Train Accuracy: 0.9174, Validation Accuracy: 0.9286, Loss: 0.1506
Epoch  58 Batch  900/1077 - Train Accuracy: 0.9309, Validation Accuracy: 0.9261, Loss: 0.1259
Epoch  58 Batch  950/1077 - Train Accuracy: 0.9356, Validation Accuracy: 0.9165, Loss: 0.1202
Epoch  58 Batch 1000/1077 - Train Accuracy: 0.9010, Validation Accuracy: 0.8991, Loss: 0.1147
Epoch  58 Batch 1050/1077 - Train Accuracy: 0.9586, Validation Accuracy: 0.9205, Loss: 0.1078
Epoch  59 Batch   50/1077 - Train Accuracy: 0.9586, Validation Accuracy: 0.9158, Loss: 0.1157
Epoch  59 Batch  100/1077 - Train Accuracy: 0.9187, Validation Accuracy: 0.9261, Loss: 0.1102
Epoch  59 Batch  150/1077 - Train Accuracy: 0.9289, Validation Accuracy: 0.9134, Loss: 0.1194
Epoch  59 Batch  200/1077 - Train Accuracy: 0.9273, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [14]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [15]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [18]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    r = [vocab_to_int[w] if w in vocab_to_int else vocab_to_int['<UNK>'] for w in sentence.split()]
    return r


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [20]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input_data:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [26, 51, 81, 212, 109, 131, 116]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [287, 216, 14, 302, 305, 229, 203, 49, 1]
  French Words: il a vu un vieux camion noir . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.